File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/Graph.java
package com.spvessel.spacevil;

import java.awt.Color;
import java.util.List;

import com.spvessel.spacevil.Flags.SizePolicy;

/**
 * Graph is class representing custom graphs with lines and points.
 * <p>
 * Contains com.spvessel.spacevil.PointsContainer and
 * com.spvessel.spacevil.LinesContainer.
 * <p>
 * Supports all events except drag and drop.
 */
public class Graph extends Prototype {
    private static int count = 0;

    PointsContainer points;
    LinesContainer lines;

    /**
     * Property to enable or disable mouse events (hover, click, press, scroll).
     * <p>
     * True: Graph can receive mouse events. False: cannot receive mouse events.
     * <p>
     * Default: False.
     */
    public boolean isHover = false;

    @Override
    protected boolean getHoverVerification(float xpos, float ypos) {
        if (isHover)
            return super.getHoverVerification(xpos, ypos);
        return false;
    }

    /**
     * Default Graph constructor. The ability to get focus is disabled by default.
     */
    public Graph() {
        setItemName("Graph_" + count);
        setBackground(0, 0, 0, 0);
        count++;

        points = new PointsContainer();
        lines = new LinesContainer();

        isFocusable = false;
    }

    /**
     * Constructs Graph with the ability to enable or disable mouse events.
     * 
     * @param hover True: Graph can receive mouse events. False: cannot receive
     *              mouse events.
     */
    public Graph(boolean hover) {
        this();
        isHover = hover;
    }

    /**
     * Initializing all elements in the Graph.
     * <p>
     * Notice: This method is mainly for overriding only. SpaceVIL calls this method
     * if necessary and no need to call it manually.
     */
    @Override
    public void initElements() {
        lines.setSizePolicy(SizePolicy.EXPAND, SizePolicy.EXPAND);
        points.setSizePolicy(SizePolicy.EXPAND, SizePolicy.EXPAND);
        addItems(lines, points);
    }

    /**
     * Setting thickness of Graph points.
     * 
     * @param thickness Point thickness.
     */
    public void setPointThickness(float thickness) {
        points.setPointThickness(thickness);
    }

    /**
     * Getting Graph points thickness.
     * 
     * @return Point thickness.
     */
    public float getPointThickness() {
        return points.getPointThickness();
    }

    /**
     * Setting Graph points color. Default: WHITE.
     * 
     * @param color Points color.
     */
    public void setPointColor(Color color) {
        points.setPointColor(color);
    }

    /**
     * Getting Graph points color.
     * 
     * @return Points color.
     */
    public Color getPointColor() {
        return points.getPointColor();
    }

    /**
     * Setting custom shape for points (if one want to use other shape than circle).
     * 
     * @return Points list of the shape as List of float[2] array.
     */
    public void setPointShape(List<float[]> shape) {
        points.setPointShape(shape);
    }

    /**
     * Getting Graph points coordinates.
     * 
     * @return Points list as List of float[2] array.
     */
    public List<float[]> getPointShape() {
        return points.getPointShape();
    }

    /**
     * Setting Graph points coordinates.
     * 
     * @return Points list as List of float[2] array.
     */
    public void setPointsCoord(List<float[]> coord) {
        points.setPoints(coord);
        lines.setPoints(coord);
    }

    /**
     * Getting Graph points coordinates.
     * 
     * @return Points list as List of float[2] array.
     */
    public List<float[]> getPointsCoord() {
        return points.getTriangles();
    }

    /**
     * Setting Graph thickness of lines.
     * 
     * @return Line thickness.
     */
    public void setLineThickness(float thickness) {
        lines.setLineThickness(thickness);
    }

    /**
     * Getting Graph lines thickness.
     * 
     * @return Lines thickness.
     */
    public float getLineThickness() {
        return lines.getLineThickness();
    }

    /**
     * Setting Graph lines color.
     * 
     * @return Line color.
     */
    public void setLineColor(Color color) {
        lines.setLineColor(color);
    }

    /**
     * Getting Graph lines color.
     * 
     * @return Lines color.
     */
    public Color getLineColor() {
        return lines.getLineColor();
    }
}

File: CSharp/SpaceVIL/SpaceVIL/Items/Graph.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using SpaceVIL.Core;

namespace SpaceVIL
{
    /// <summary>
    /// Graph is class representing custom graphs with lines and points.
    /// <para/> Contains SpaceVIL.PointsContainer and SpaceVIL.LinesContainer.
    /// <para/> Supports all events except drag and drop.
    /// </summary>
    public class Graph : Prototype
    {
        private static int count = 0;

        private PointsContainer points;
        private LinesContainer lines;
        /// <summary>
        /// Property to enable or disable mouse events (hover, click, press, scroll).
        /// <para/> True: Graph can receive mouse events. False: cannot receive mouse events.
        /// <para/> Default: False.
        /// </summary>
        public bool IsHover = false;

        protected internal override bool GetHoverVerification(float xpos, float ypos)
        {
            if (IsHover)
                return base.GetHoverVerification(xpos, ypos);
            return false;
        }
        /// <summary>
        /// Default Graph constructor. The ability to get focus is disabled by default.
        /// </summary>
        public Graph()
        {
            SetItemName("Graph_" + count);
            SetBackground(Color.Transparent);
            count++;

            points = new PointsContainer();
            lines = new LinesContainer();

            IsFocusable = false;
        }
        /// <summary>
        /// Constructs Graph with the ability to enable or disable mouse events.
        /// </summary>
        /// <param name="hover">True: Graph can receive mouse events. 
        /// False: cannot receive mouse events.</param>
        public Graph(bool hover) : this()
        {
            IsHover = hover;
        }
        /// <summary>
        /// Initializing all elements of Graph.
        /// <para/> Notice: This method is mainly for overriding only. SpaceVIL calls 
        /// this method if necessary and no need to call it manually.
        /// </summary>
        public override void InitElements()
        {
            lines.SetSizePolicy(SizePolicy.Expand, SizePolicy.Expand);
            points.SetSizePolicy(SizePolicy.Expand, SizePolicy.Expand);
            AddItems(lines, points);
        }
        /// <summary>
        /// Setting thickness of Graph points.
        /// </summary>
        /// <param name="thickness">Point thickness.</param>
        public void SetPointThickness(float thickness)
        {
            points.SetPointThickness(thickness);
        }
        /// <summary>
        /// Getting Graph points thickness.
        /// </summary>
        /// <returns>Point thickness.</returns>
        public float GetPointThickness()
        {
            return points.GetPointThickness();
        }
        /// <summary>
        /// Setting Graph points color. Default: White.
        /// </summary>
        /// <param name="color">Points color.</param>
        public void SetPointColor(Color color)
        {
            points.SetPointColor(color);
        }
        /// <summary>
        /// Getting Graph points color.
        /// </summary>
        /// <returns>Points color.</returns>
        public Color GetPointColor()
        {
            return points.GetPointColor();
        }
        /// <summary>
        /// Setting custom shape for points (if one want to use other shape than circle).
        /// </summary>
        /// <param name="shape">Points list of the shape as List of float[2] array.</param>
        public void SetPointShape(List<float[]> shape)
        {
            points.SetPointShape(shape);
        }
        /// <summary>
        /// Getting current shape of points. Default: circle shape.
        /// </summary>
        /// <returns>Points list as List of float[2] array.</returns>
        public List<float[]> GetPointShape()
        {
            return points.GetPointShape();
        }
        /// <summary>
        /// Setting Graph points coordinates.
        /// </summary>
        /// <param name="coord">Points list as List of float[2] array.</param>
        public void SetPoints(List<float[]> coord)
        {
            points.SetPoints(coord);
            lines.SetPoints(coord);
        }
        /// <summary>
        /// Getting Graph points coordinates.
        /// </summary>
        /// <returns>Points list as List of float[2] array.</returns>
        public List<float[]> GetPoints()
        {
            return points.GetTriangles();
        }
        /// <summary>
        /// Setting Graph thickness of lines.
        /// </summary>
        /// <param name="thickness">Line thickness.</param>
        public void SetLineThickness(float thickness)
        {
            lines.SetLineThickness(thickness);
        }
        /// <summary>
        /// Getting Graph lines thickness.
        /// </summary>
        /// <returns>Lines thickness.</returns>
        public float GetLineThickness()
        {
            return lines.GetLineThickness();
        }
        /// <summary>
        /// Setting Graph lines color.
        /// </summary>
        /// <param name="color">Line color.</param>
        public void SetLineColor(Color color)
        {
            lines.SetLineColor(color);
        }
        /// <summary>
        /// Getting Graph lines color.
        /// </summary>
        /// <returns>Lines color.</returns>
        public Color GetLineColor()
        {
            return lines.GetLineColor();
        }
    }
}